### PR TITLE
libhns: Code optimization and bugfixes for libhns

### PR DESCRIPTION
**Pull request by oulijun**
Here is optimize some codes for hns userspace drvier as well as fixes a bug with nvme application。

### CODE REVIEW

File: providers/hns/hns_roce_u_db.h [hunk 1]
```diff
@@ -38,9 +38,9 @@
 #define _HNS_ROCE_U_DB_H
 
 #if __BYTE_ORDER == __LITTLE_ENDIAN
-#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) val[1] << 32 | val[0])
+#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) (val[1]) << 32 | (val[0]))
 #elif __BYTE_ORDER == __BIG_ENDIAN
-#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) val[0] << 32 | val[1])
+#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) (val[0]) << 32 | (val[1]))
 #else
```
Comment (jgunthorpe) [inline providers/hns/hns_roce_u_db.h]: Please use one of the mmio_write64's instead of this mess

Comment (jgunthorpe) [reply providers/hns/hns_roce_u_db.h]: Also the placement of the brackets is wrong

File: providers/hns/hns_roce_u_hw_v2.c [hunk 2]
```diff
@@ -109,6 +109,9 @@ static void hns_roce_v2_handle_error_cqe(struct hns_roce_v2_cqe *cqe,
 		wc->status = IBV_WC_GENERAL_ERR;
 		break;
 	}
+
+	if (wc->status != IBV_WC_WR_FLUSH_ERR)
+		fprintf(stderr, PFX "error cqe status: 0x%x\n", cqe_status);
```
Comment (jgunthorpe) [inline providers/hns/hns_roce_u_hw_v2.c]: No prints from providers.

**Comment by oulijun** [comment]
在 2019/4/23 23:53, Jason Gunthorpe 写道:
>
> *@jgunthorpe* requested changes on this pull request.
>
> ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
>
> In providers/hns/hns_roce_u_db.h <https://github.com/linux-rdma/rdma-core/pull/511#discussion_r277729982>:
>
> >  #elif __BYTE_ORDER == __BIG_ENDIAN
> -#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) val[0] << 32 | val[1])
> +#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) (val[0]) << 32 | (val[1]))
>  #else
>
> Please use one of the mmio_write64's instead of this mess
>
Good advice. I could generate the followup patch for insteading it.
> ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
>
> In providers/hns/hns_roce_u_db.h <https://github.com/linux-rdma/rdma-core/pull/511#discussion_r277730139>:
>
> >  #elif __BYTE_ORDER == __BIG_ENDIAN
> -#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) val[0] << 32 | val[1])
> +#define HNS_ROCE_PAIR_TO_64(val) ((uint64_t) (val[0]) << 32 | (val[1]))
>  #else
>
> Also the placement of the brackets is wrong
>
> ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
>
> In providers/hns/hns_roce_u_hw_v2.c <https://github.com/linux-rdma/rdma-core/pull/511#discussion_r277730269>:
>
> > @@ -109,6 +109,9 @@ static void hns_roce_v2_handle_error_cqe(struct hns_roce_v2_cqe *cqe,
>  		wc->status = IBV_WC_GENERAL_ERR;
>  		break;
>  	}
> +
> +	if (wc->status != IBV_WC_WR_FLUSH_ERR)
> +		fprintf(stderr, PFX "error cqe status: 0x%x\n", cqe_status);
>
> No prints from providers.
>
Ok, I will fix it.

File: providers/hns/hns_roce_u_hw_v2.c [hunk 3]
```diff
@@ -404,6 +407,11 @@ static int hns_roce_handle_recv_inl_wqe(struct hns_roce_v2_cqe *cqe,
 		sge_list = (*cur_qp)->rq_rinl_buf.wqe_list[wr_cnt].sg_list;
 		sge_num = (*cur_qp)->rq_rinl_buf.wqe_list[wr_cnt].sge_cnt;
 		wqe_buf = (uint8_t *)get_recv_wqe_v2(*cur_qp, wr_cnt);
+		if (!wqe_buf) {
+			fprintf(stderr,	PFX "can't get recv inl wqe !\n");
```
Comment (rleon) [inline providers/hns/hns_roce_u_hw_v2.c]: You was asked to do not add prints to library.

File: providers/hns/hns_roce_u_hw_v2.c [hunk 4]
```diff
@@ -959,18 +968,24 @@ static int hns_roce_u_v2_post_recv(struct ibv_qp *ibvqp, struct ibv_recv_wr *wr,
...
 		if (wr->num_sge > qp->rq.max_gs) {
-			ret = -1;
+			ret = -EINVAL;
 			*bad_wr = wr;
 			goto out;
 		}
 
 		wqe = get_recv_wqe_v2(qp, ind);
+		if (!wqe) {
+			fprintf(stderr,	PFX "can't get recv wqe !\n");
```
Comment (rleon) [inline providers/hns/hns_roce_u_hw_v2.c]: Same

**Comment by oulijun** [comment]
Hi, Leon & Jason
    The travisCI have a new bug? I have check the log and don't find the build error.

Thanks
Lijun

**Comment by rleon** [comment]
I restarted it for you, TravisCI failed because of our CI job was too long.

**Comment by nmorey** [comment]
There seems to be an issue with the SUSE mirror used for downloading the packages.
I'll try pinging some people today to see if they know what's going on...

**Pull request merged**